### PR TITLE
Add new API PresignedHeadObject method.

### PR DESCRIPTION
**Pull request by harshavardhana**
This PR borrows original work from @dmage in PR #770

### CODE REVIEW

File: api-presigned.go [hunk 1]
```diff
@@ -91,12 +91,24 @@ func (c Client) PresignedGetObject(bucketName string, objectName string, expires
...
+}
+
 // PresignedPutObject - Returns a presigned URL to upload an object without credentials.
 // Expires maximum is 7days - ie. 604800 and minimum is 1.
 func (c Client) PresignedPutObject(bucketName string, objectName string, expires time.Duration) (u *url.URL, err error) {
 	return c.presignURL("PUT", bucketName, objectName, expires, nil)
 }
 
+// Presign - returns a presigned URL for any http method of your choice.
+// Expires maximum in 7days - i.e 604800 and minimum in 1.
```
Comment (poornas) [inline api-presigned.go]: Expires maximum in 7days  -> Expires maximum is 7 days.

File: docs/API.md [hunk 2]
```diff
@@ -1093,14 +1091,44 @@ if err != nil {
 fmt.Println(presignedURL)
 ```
 
+<a name="PresignedHeadObject"></a>
+### PresignedHeadObject(bucketName, objectName string, expiry time.Duration, reqParams url.Values) (*url.URL, error)
+
+Generates a presigned URL for HTTP HEAD operations. Browsers/Mobile clients may point to this URL to directly get metadata from objects even if the bucket is private. This presigned URL can have an associated expiration time in seconds after which it is no longer operational. The default expiry is set to 7 days.
+
```
[the remark this answers is not in the document]
Comment (harshavardhana) [reply docs/API.md]: No default is 7 days if not set that is.

File: functional_tests.go [hunk 3]
```diff
@@ -2166,14 +2166,32 @@ func testFunctional() {
 		logger().Fatal("Error: ", err)
 	}
 
+	// Generate presigned HEAD object url.
+	presignedHeadURL, err := c.PresignedHeadObject(bucketName, objectName, 3600*time.Second, nil)
```
Comment (vadmeste) [inline functional_tests.go]: Can you please use another instance of Minio client initialized without credentials ? We will be more close to real world use case

File: examples/s3/presignedheadobject.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+func main() {
+	// Note: YOUR-ACCESSKEYID, YOUR-SECRETACCESSKEY, my-bucketname and my-objectname
+	// are dummy values, please replace them with original values.
+
+	// Requests are always secure (HTTPS) by default. Set secure=false to enable insecure (HTTP) access.
+	// This boolean value is the last argument for New().
+
+	// New returns an Amazon S3 compatible client object. API compatibility (v2 or v4) is automatically
+	// determined based on the Endpoint value.
+	s3Client, err := minio.New("s3.amazonaws.com", "YOUR-ACCESSKEYID", "YOUR-SECRETACCESSKEY", true)
```
Comment (vadmeste) [inline examples/s3/presignedheadobject.go]: This example should be updated to do not use access and secret keys

File: docs/API.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (vadmeste) [reply docs/API.md]: But also the maximum is 7 AFAIK

**Comment by harshavardhana** [comment]
Can you guys review again @vadmeste @poornas ?

File: api-presigned.go [hunk 5]
```diff
@@ -84,25 +84,40 @@ func (c Client) presignURL(method string, bucketName string, objectName string,
...
+// the query parameters.
 func (c Client) PresignedHeadObject(bucketName string, objectName string, expires time.Duration, reqParams url.Values) (u *url.URL, err error) {
 	return c.presignURL("HEAD", bucketName, objectName, expires, reqParams)
 }
 
-// PresignedPutObject - Returns a presigned URL to upload an object without credentials.
-// Expires maximum is 7days - ie. 604800 and minimum is 1.
+// PresignedPutObject - Returns a presigned URL to upload an object
+// without credentials. URL can have a maximum expiry of upto 7days
+// or a minimum of 1sec, if expiry duration is set to 0 URL a
```
Comment (vadmeste) [inline api-presigned.go]: I didn't test it but after reading the code, it seems setting 0 to expires returns an error (invoked by isValidExpiry() in presignURL())

Comment (harshavardhana) [reply api-presigned.go]: I see looks like behavior has changed.. @vadmeste - let me change the comment to reflect the new changes.

File: api-presigned.go [hunk 6]
```diff
@@ -84,19 +84,40 @@ func (c Client) presignURL(method string, bucketName string, objectName string,
 }
 
 // PresignedGetObject - Returns a presigned URL to access an object
-// without credentials. Expires maximum is 7days - ie. 604800 and
-// minimum is 1. Additionally you can override a set of response
-// headers using the query parameters.
+// data without credentials. URL can have a maximum expiry of
+// upto 7days or a minimum of 1sec, if expiry duration is set to 0
+// URL a maximum value of 7 days is automatically choosen. Additionally
```
Comment (poornas) [inline api-presigned.go]: +// upto 7days or a minimum of 1sec, if expiry duration is set to 0
 +// URL a maximum value of 7 days is automatically choosen.

Should be:
+// upto 7days or a minimum of 1sec, if expiry duration is set to 0
 +// ~~URL~~ a maximum value of 7 days is automatically cho~~o~~sen.

Ditto for the other 3 function header comments

**Comment by harshavardhana** [comment]
@vadmeste waiting on your review.

**Pull request merged**